packages/flutter/lib/src/rendering/sliver_list.dart

```diff
@@ -38,24 +47,171 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
+  }
+
+  bool _canUseCachedExtent(int index) {
+    assert(enableItemExtentsCaching);
+    if (_cachedItemExtents.containsKey(index)) {
+      final Key? currentKey = childManager.keyOf(index);
+      if (_cachedItemExtents[index]!.childKey == currentKey) {
+        return true;
+      } else {
+        // Remove the cached extent if widget's key changed.
+        _cachedItemExtents.remove(index);
+        return false;
+      }
+    }
+    return false;
+  }
+
+  RenderBox? _obtainOneLeadingChild() {
+    final int leadingIndex = indexOf(firstChild!) - 1;
+    final bool layoutImmediate = !enableItemExtentsCaching || !_canUseCachedExtent(leadingIndex);
+    final RenderBox? leadingChild = insertAndLayoutLeadingChild(
+      _childConstraints,
+      parentUsesSize: true,
+      layoutImmediate: layoutImmediate,
+    );
+    if (enableItemExtentsCaching && leadingChild != null) {
+      assert(leadingChild == firstChild);
+      if (layoutImmediate) {
+        _cacheItemExtent(firstChild!);
+      } else {
+        _childrenWithoutLayout.add(leadingIndex);
+      }
+    }
+    return leadingChild;
+  }
+
+  RenderBox? _obtainOneTrailingChild(RenderBox after) {
+    final int trailingIndex = indexOf(after) + 1;
+    final bool layoutImmediate = !enableItemExtentsCaching || !_canUseCachedExtent(trailingIndex);
+    final RenderBox? trailingChild = insertAndLayoutChild(
+      _childConstraints,
+      after: after,
+      parentUsesSize: true,
+      layoutImmediate: layoutImmediate,
+    );
+    if (enableItemExtentsCaching && trailingChild != null) {
+      if (layoutImmediate) {
+        _cacheItemExtent(trailingChild);
+      } else {
+        _childrenWithoutLayout.add(trailingIndex);
+      }
+    }
+    return trailingChild;
+  }
+
+  void _processChildLayout(RenderBox child) {
+    final int childIndex = indexOf(child);
+    final bool layoutImmediate = !enableItemExtentsCaching || !_canUseCachedExtent(childIndex);
+    if (layoutImmediate) {
+      child.layout(_childConstraints, parentUsesSize: true);
+      if (enableItemExtentsCaching) {
+        _cacheItemExtent(child);
+      }
+    } else {
+      _childrenWithoutLayout.add(childIndex);
+    }
+  }
+
+  void _performLoadingScopeChildrenLayoutIfNeeded() {
+    if (!enableItemExtentsCaching) {
+      return;
+    }
+    for (RenderBox? child = firstChild; child != null ; child = childAfter(child)) {
+      final int index = indexOf(child);
+      if (_childrenWithoutLayout.contains(index)) {
+        child.layout(_childConstraints, parentUsesSize: true);
+        // Throw if the real extent differs from the cached value.
+        assert(() {
+          if (super.paintExtentOf(child!) != _cachedItemExtents[index]!.mainAxisExtent) {
+            throw FlutterError(
+              'The item(index: $index) main axis extent has changed after caching, if this is '
+              'expected, change its Key to update the cached value.',
```

Comment:
```Changing the key might not be an option in some cases.```
You are right, we should provide a way to invalidate certain caches.

```If the extent is different, can we just update it?```
Maybe we should throw here because we've used the wrong cached value to find the target position.

Reply:
This seems dangerous if developers are not sure what they are doing. It is also a bit confusing that it would require manual invalidation of caches. I think what would happen is developer will just use the invalidating caches everywhere so that it doesn't crash.